nautobot/core/views/drf_views.py

```diff
@@ -0,0 +1,835 @@
+    "update": "change",
+    "bulk_create": "add",
+    "bulk_destroy": "delete",
+    "bulk_update": "change",
+}
+
+
+class NautobotViewSetMixin(ViewSetMixin, generics.GenericAPIView, AccessMixin, GetReturnURLMixin, FormView):
+    """
+    serializer_class has to be specified to eliminate the need to override retrieve() in the RetrieveModelMixin for now.
+    It is a step forward in our transition from NetBox legacy code to DRF framework.
+    NautobotHTMLRenderer is inherited from BrowsableAPIRenderer to render the original context needed for rendering the templates.
+    """
+
+    serializer_class = None
+    renderer_classes = [NautobotHTMLRenderer]
+
+    def get_permissions_for_model(self, model, actions):
+        """
+        Resolve the named permissions for a given model (or instance) and a list of actions (e.g. view or add).
+
+        :param model: A model or instance
+        :param actions: A list of actions to perform on the model
+        """
+        permissions = []
+        for action in actions:
+            if action not in ("view", "add", "change", "delete"):
+                raise ValueError(f"Unsupported action: {action}")
+        permissions.append("{}.{}_{}".format(model._meta.app_label, action, model._meta.model_name))
+        return permissions
+
+    def get_required_permission(self):
+        """
+        Obtain the permissions needed to perform certain actions on a model.
+        """
+        return self.get_permissions_for_model(self.queryset.model, [PERMISSIONS_ACTION_MAP[self.action]])
+
+    def has_permission(self):
+        """
+        Check whether the user has the permissions needed to perform certain actions.
+        """
+        user = self.request.user
+        permission_required = self.get_required_permission()
+        # Check that the user has been granted the required permission(s).
+        if user.has_perms(permission_required):
+
+            # Update the view's QuerySet to filter only the permitted objects
+            for permission in permission_required:
+                action = resolve_permission(permission)[1]
+                self.queryset = self.queryset.restrict(user, action)
+
+            return True
+
+        return False
+
+    def check_permissions(self, request):
+        """
+        Used to determine whether the user has permissions to a view.
+        Using AccessMixin handle_no_permission() to deal with Object-Level permissions and API-Level permissions in one pass.
+        """
+        if not self.has_permission():
+            self.handle_no_permission()
+
+    def check_object_permissions(self, request, obj):
+        """
+        Used to determine whether the user has the object-level permissions to perform certain actions to a model instance.
+        Using AccessMixin handle_no_permission() to deal with Object-Level permissions and API-Level permissions in one pass.
+        """
+        if not self.has_permission():
+            self.handle_no_permission()
```

Comment:
check_permissions() is called everytime in this [initial() functioon](https://github.com/encode/django-rest-framework/blob/791d48c6900e758898cd10ee2db45a76c200522a/rest_framework/views.py#L399), check_object_permission is called in this line. Since we do call get_object() everytime we have a request, this will result in checking the permission twice. I wonder if we can write our own get_object() function that dont call the function once check_permissions() passes.

Reply:
Yeah `check_object_permissions()` can be removed from `get_object()` since `check_permission()` has object-level permissions covered